### PR TITLE
Re-designing credentials, adding x509

### PR DESCRIPTION
**Pull request by RLeclair**
PR contains design changes for az_credentials. Includes x509 credential support.

### CODE REVIEW

**Review by CIPop — APPROVED** [review]

File: sdk/inc/azure/core/az_credentials_common.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * @brief Pointer to a credential implementation.
+ */
+typedef void* az_platform_credential;
+
+/**
+ * @brief Defines the supported credential types.
+ */
+typedef enum
```
Comment (RickWinter) [inline sdk/inc/azure/core/az_credentials_common.h:44]: This should be removed.  Core does not know about credential types it only knows about the AccessToken and the TokenCredential type.  Types expanding on TokenCredential are identity constructs.

**Review by RickWinter — CHANGES_REQUESTED** [review]

File: sdk/inc/azure/core/az_credentials_common.h (hunk 1 above)
Comment (RickWinter) [reply sdk/inc/azure/core/az_credentials_common.h:44]: You can use the C++ impl to reference the Azure SDK design patterns for credentials.
https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp

Comment (CIPop) [reply sdk/inc/azure/core/az_credentials_common.h:44]: The change is based from an offline discussion we had with @LarryOsterman . Sharing the summary here:

> We will create pal_x509certifiate which represents the certificate for the selected platform (e.g. MosquittoMQTT).
> - az_credential  will become a public type.  
> - We will add a constructor (_init function) that accepts a pal_x509certificate 
> - A separate constructor will be used for SAS az_credentials

Comment (CIPop) [reply sdk/inc/azure/core/az_credentials_common.h:44]: @RickWinter what would be the recommended way to add a public API for X.509 credentials?
Should we create our own `az_iot_` type instead?

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#include <azure/core/_az_cfg_prefix.h>
+
+/**
+ * @brief Defines the key storage types.
+ */
+typedef enum
+{
+  AZ_CREDENTIALS_X509_KEY_MEMORY = 0,
+  AZ_CREDENTIALS_X509_KEY_SECURITY_MODULE = 1,
+} az_platform_credential_x509_key_type;
```
Comment (ewertons) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:33]: Maybe a name like `az_platform_credential_x509_key_source` would be more appropriate?

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  struct
+  {
+    /// Contains x509 certificate
+    az_span cert;
+
+    /// Contains x509 key
+    az_span key;
+
+    /// Specifies the storage type for the key.
+    az_platform_credential_x509_key_type key_type;
```
Comment (ewertons) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:49]: Here too, key type relates to the type of the key itself. 
The concept being considered here is: where is this key coming from?
So maybe `storage_source` or `key_source` would be more correct.

**Review by ewertons — CHANGES_REQUESTED** [review]

File: sdk/inc/azure/platform/az_platform_x509_credential.h (hunk 2 above)
Comment (ericwol-msft) [reply sdk/inc/azure/platform/az_platform_x509_credential.h:33]: also 
AZ_CREDENTIALS_X509_KEY_STORAGE_MEMORY
AZ_CREDENTIALS_X509_KEY_STORAGE_SECURITY_MODULE
?

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * @brief x509 credential definition.
+ */
+typedef struct
+{
+  struct
+  {
+    /// Contains x509 certificate
+    az_span cert;
```
Comment (ericwol-msft) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:43]:
```suggestion
    az_span certificate;
```

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// Contains x509 key
```
Comment (ericwol-msft) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:45]:
```suggestion
    /// Contains x509 private key
```

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    az_span key;
+
+    /// Specifies the storage type for the key.
+    az_platform_credential_x509_key_type key_type;
+  } _internal;
+} az_platform_x509_credential;
```
Comment (ericwol-msft) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:51]: So how is this different from a ECC cert?

File: sdk/inc/azure/platform/az_platform_x509_credential.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * @brief x509 credential constructor.
+ */
+AZ_NODISCARD AZ_INLINE az_credential az_credential_x509_init(az_platform_x509_credential* cred)
```
Comment (ericwol-msft) [inline sdk/inc/azure/platform/az_platform_x509_credential.h:56]:
```suggestion
AZ_NODISCARD AZ_INLINE az_credential az_credential_x509_init(az_platform_x509_certificate* certificate)
```

Comment (ericwol-msft) [reply sdk/inc/azure/platform/az_platform_x509_credential.h:56]: specific credential in (cert), generic credential out.

File: sdk/inc/azure/platform/az_platform_x509_credential.h (hunk 3 above)
Comment (RLeclair) [reply sdk/inc/azure/platform/az_platform_x509_credential.h]: Changed to `key_source` thanks for pointing it out!

File: sdk/inc/azure/platform/az_platform_x509_credential.h (hunk 2 above)
Comment (RLeclair) [reply sdk/inc/azure/platform/az_platform_x509_credential.h]: Added both changes, thanks!

File: sdk/inc/azure/core/az_credentials_common.h (hunk 1 above)
Comment (RLeclair) [reply sdk/inc/azure/core/az_credentials_common.h]: Hi @RickWinter, I've gone ahead and removed `az_credentials_common.h` and moved my changes to `az_iot_certificate_x509`.

**Comment by RLeclair** [comment]
Updated the location of the credential type. Changed the naming to `az_iot_certificate_x509.h` and removed the abstraction layer.

**Pull request closed without merging**